### PR TITLE
(WIP) fix: Facet pivots don't support limits

### PR DESCRIPTION
**Pull request by tsterker**
**Update** (2021-04-04): I realized this feature was explicitly added in https://github.com/solariumphp/solarium/pull/732 and pointed out as missing in https://github.com/solariumphp/solarium/issues/713.
Now I wonder if I am missing something, but for me the [Solr documentation](https://solr.apache.org/guide/8_6/faceting.html#additional-pivot-parameters) seemed rather clear and also [google](https://www.google.com/search?q=%22facet.pivot.limit%22) does not really yield anything for `facet.pivot.limit`.

---

> ⚠️  This PR is **work-in-progress**:
> - Did not run/update any tests
> - Not sure if I might be missing something in the Solr specification

I just hunted down a bug in my app where I did **not properly set the facet limits**.

Calling `setLimit` on a `Pivot` results in a `facet.pivot.limit` parameter to be set in the Solr request, which does not seem to be a supported parameter. Instead, the limit must be set on the `FacetSet` itself (which will yield `facet.limit`). See [Solr documentation](https://solr.apache.org/guide/8_6/faceting.html#additional-pivot-parameters) with a reference to this.

Unfortunately, I currently can't spend more time on this topic (investigation/implementation), but wanted to already start a conversation with this PR.

### CODE REVIEW

**Comment by thomascorthals** [comment]
> ### Additional Pivot Parameters
> 
> Although `facet.pivot.mincount` deviates in name from the `facet.mincount` parameter used by field faceting, many of the faceting parameters described above can also be used with pivot faceting:
> 
> - `facet.limit`
> - `facet.offset`
> - `facet.sort`
> - `facet.overrequest.count`
> - `facet.overrequest.ratio` 

This does seem to say that `facet.pivot.mincount` is the only parameter with a  `.pivot` tmesis. It could be written a bit more clearly though, and would benefit from an example.

One place to find working examples is in Solr's own test suite. No mention of `facet.pivot.limit` anywhere in the tests or the source code. But they do combine `facet.pivot` with `facet.limit` in the tests.

https://github.com/apache/solr/blob/5bb9ecdb87e21dc912c52945fe027ccc63b47e0a/solr/core/src/test/org/apache/solr/handler/component/DistributedFacetPivotLongTailTest.java#L73-L77

```java
    SolrParams req = params( "q", "*:*", 
                             "distrib", "false",
                             "facet", "true", 
                             "facet.limit", "10",
                             "facet.pivot", "foo_s,bar_s");
```

https://github.com/apache/solr/blob/5bb9ecdb87e21dc912c52945fe027ccc63b47e0a/solr/core/src/test/org/apache/solr/handler/component/DistributedFacetPivotSmallAdvancedTest.java#L114-L116

```java
    facetParams.add(params("facet", "true",
                           "facet.limit", "1",
                           "facet.pivot", "{!stats=s1}place_t,company_t"));
```

https://github.com/apache/solr/blob/5bb9ecdb87e21dc912c52945fe027ccc63b47e0a/solr/core/src/test/org/apache/solr/handler/component/FacetPivotSmallTest.java#L402-L409

```java
      SolrParams p = SolrParams.wrapDefaults(params("q", "*:*",
              "rows", "0",
              "facet", "true",
              "facet.pivot", "company_t",
              "facet.sort", "index",
              "facet.pivot.mincount", "4",
              "facet.limit", "4"),
          variableParams);
```

**Comment by tsterker** [comment]
@thomascorthals Thanks for your details.

Do I understand correctly that you would agree that there is no such thing as `facet.pivot.limit`? This would further supported by me failing to get expected results with `facet.pivot.limit` when directly interacting with Solr.

**Comment by thomascorthals** [comment]
> Do I understand correctly that you would agree that there is no such thing as `facet.pivot.limit`? This would further supported by me failing to get expected results with `facet.pivot.limit` when directly interacting with Solr.

I do agree that there is no such thing as `facet.pivot.limit`.

The `setLimit()`/`getLimit()` methods don't need to be removed though. The parameter should simply be `facet.limit`.

File: src/Component/Facet/Pivot.php [hunk 1]
```diff
@@ -36,28 +36,6 @@ public function getType(): string
...
-    /**
-     * Get the facet limit.
-     *
-     * @return int
-     */
-    public function getLimit(): ?int
-    {
-        return $this->getOption('limit');
-    }
-
```
Comment (thomascorthals) [inline src/Component/Facet/Pivot.php:60]: These methods can stay!

File: src/Component/RequestBuilder/FacetSet.php [hunk 2]
```diff
@@ -282,7 +282,6 @@ public function addFacetPivot($request, $facet): void
             )
         );
         $request->addParam('facet.pivot.mincount', $facet->getMinCount(), true);
-        $request->addParam('facet.pivot.limit', $facet->getLimit(), true);
```
Comment (thomascorthals) [inline src/Component/RequestBuilder/FacetSet.php:285]:
```suggestion
        $request->addParam('facet.limit', $facet->getLimit(), true);
```

**Review by thomascorthals — CHANGES_REQUESTED** [review]
You'll have to update the unit test as well.

https://github.com/solariumphp/solarium/blob/fa93d0d5fb5b6797922f4e486e957d789c2bfba3/tests/Component/RequestBuilder/FacetSetTest.php#L374-L377

**Comment by thomascorthals** [comment]
@tsterker Looking into this further, it turned out there were bigger issues with pivot facets than this misnamed parameter. I've opened PR #963 to address them all at once. You're welcome to join the discussion there!

**Pull request closed without merging**